### ISSUE TITLE
Add support for garuda linux

### ISSUE THREAD

**Issue by jstelzer** [topgrade-rs/topgrade#226]
### Which operating systems are supported by this tool?

Garuda linux is arch plus some patches and update tool.
It is recommended to use their update tool so that all customization's are kept in synch.

### What should Topgrade do to figure out if the tool needs to be invoked?

If `garuda-update` is in $PATH, use that instead of directly invoking pacman/paru/powerpill
This script does some additional work and should be used instead of a partial update.

### Which exact commands should Topgrade run?
```
garuda-update
```

## More information
I implemented this minor patch here:
https://github.com/jstelzer/topgrade/tree/add-garuda-update-support

I'm not a proficient rust developer, but i'm trying to learn. If you're up for it i will open a PR with my patch.
My biggest question is path precedence order/topgrade deciding on which arch upgrade path/struct to use. Right now the GarudaUpdate impl is 'first'. On my system this is the (correct from my perspective) output of `topgrade -n`

```
❯ topgrade -n

―― 10:17:24 - System update ――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/garuda-update

―― 10:17:24 - Configuration update ―――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/sudo '--preserve-env=DIFFPROG' /usr/bin/pacdiff

―― 10:17:24 - oh-my-zsh ――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Pulling custom plugins and themes
Would pull /home/mental/.oh-my-zsh/custom/themes/powerlevel10k
Dry running: zsh /home/mental/.oh-my-zsh/tools/upgrade.sh

―― 10:17:24 - TLDR ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/tldr --update

―― 10:17:24 - rustup ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /home/mental/.cargo/bin/rustup self update
Dry running: /home/mental/.cargo/bin/rustup update

―― 10:17:24 - Cargo ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /home/mental/.cargo/bin/cargo-install-update install-update --git --all

―― 10:17:24 - Emacs ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/emacs --batch --debug-init -l /home/mental/.emacs.d/init.el --eval '(when (fboundp '\''paradox-upgrade-packages)      (progn        (unless (boundp '\''paradox-github-token)          (setq paradox-github-token t))        (paradox-upgrade-packages)        (princ         (if (get-buffer "*Paradox Report*")             (with-current-buffer "*Paradox Report*" (buffer-string))           "\nNothing to upgrade\n"))))  '

―― 10:17:24 - pip3 ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/python3 -m pip install --upgrade --user pip

―― 10:17:24 - Containers ―――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/docker pull alpine:latest
Dry running: /usr/bin/docker pull postgres:13
Dry running: /usr/bin/docker pull ghcr.io/linuxserver/plex:latest
Dry running: /usr/bin/docker pull node:lts-alpine

―― 10:17:24 - micro ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/micro -plugin update

―― 10:17:24 - Firmware upgrades ――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/fwupdmgr refresh
Dry running: /usr/bin/fwupdmgr get-updates

―― 10:17:24 - Summary ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
System update: OK
config-update: OK
oh-my-zsh: OK
TLDR: OK
rustup: OK
cargo: OK
Emacs: OK
pip3: OK
Containers: OK
Firmware upgrades: OK
```

Testing the fallback logic, if i rename the `garuda-update` script I see the original old test plan:
```
❯ topgrade -n

―― 10:18:34 - System update ――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/paru --pacman /usr/bin/powerpill -Syu

―― 10:18:34 - Configuration update ―――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/sudo '--preserve-env=DIFFPROG' /usr/bin/pacdiff

―― 10:18:34 - oh-my-zsh ――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Pulling custom plugins and themes
Would pull /home/mental/.oh-my-zsh/custom/themes/powerlevel10k
Dry running: zsh /home/mental/.oh-my-zsh/tools/upgrade.sh

―― 10:18:34 - TLDR ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/tldr --update

―― 10:18:34 - rustup ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /home/mental/.cargo/bin/rustup self update
Dry running: /home/mental/.cargo/bin/rustup update

―― 10:18:34 - Cargo ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /home/mental/.cargo/bin/cargo-install-update install-update --git --all

―― 10:18:34 - Emacs ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/emacs --batch --debug-init -l /home/mental/.emacs.d/init.el --eval '(when (fboundp '\''paradox-upgrade-packages)      (progn        (unless (boundp '\''paradox-github-token)          (setq paradox-github-token t))        (paradox-upgrade-packages)        (princ         (if (get-buffer "*Paradox Report*")             (with-current-buffer "*Paradox Report*" (buffer-string))           "\nNothing to upgrade\n"))))  '

―― 10:18:34 - pip3 ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/python3 -m pip install --upgrade --user pip

―― 10:18:34 - Containers ―――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/docker pull alpine:latest
Dry running: /usr/bin/docker pull postgres:13
Dry running: /usr/bin/docker pull ghcr.io/linuxserver/plex:latest
Dry running: /usr/bin/docker pull node:lts-alpine

―― 10:18:34 - micro ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/micro -plugin update

―― 10:18:34 - Firmware upgrades ――――――――――――――――――――――――――――――――――――――――――――――――
Dry running: /usr/bin/fwupdmgr refresh
Dry running: /usr/bin/fwupdmgr get-updates

―― 10:18:34 - Summary ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
System update: OK
config-update: OK
oh-my-zsh: OK
TLDR: OK
rustup: OK
cargo: OK
Emacs: OK
pip3: OK
Containers: OK
Firmware upgrades: OK
```

**Comment by jstelzer** [comment]
why add latency? Here is the PR for your feedback/decision: https://github.com/topgrade-rs/topgrade/pull/227

**Comment by lisandrocarmona** [comment]
@jstelzer I don't want to hijack your thread, but Topgrade was working in Garuda some days ago, and now, it's not being able to update the system. Will the "garuda-update" method you're suggesting solve my issue?

Log for the following command:
`RUST_BACKTRACE=1 RUST_BACKTRACE=full topgrade -v`
https://bin.garudalinux.org/?2d97efee132b9399#AMSohhBwXL58Fz1UjWPJA9tauJWmL63ybJjSztU9VxNd

**Comment by DottoDev** [comment]
> @jstelzer I don't want to hijack your thread, but Topgrade was working in Garuda some days ago, and now, it's not being able to update the system. Will the "garuda-update" method you're suggesting solve my issue?
> 
> Log for the following command: `RUST_BACKTRACE=1 RUST_BACKTRACE=full topgrade -v` https://bin.garudalinux.org/?2d97efee132b9399#AMSohhBwXL58Fz1UjWPJA9tauJWmL63ybJjSztU9VxNd

@lisandrocarmona This is due to the error mentioned in #220 which will be fixed with #228 .

**Comment by lisandrocarmona** [comment]
Thank you @DottoDev. Now back to the original issue of integrating garuda-update from the OP.

**Comment by jstelzer** [comment]
Hey folks, glad i'm not the only one interested in this. Initially topgrade was running ok. Then at some point I hit some issues with how powerpill was configured and fixed that. It was at that point I became aware garuda-update was the recommended path for this distro. So I added an entry to the enum of arch updaters and implemented 2 callbacks. Sorry if i'm slow to respond. It's a holiday weekend in the USA and I'm out enjoying the tail end of motorcycle season.

**Comment by jstelzer** [comment]
Thanks for merging

**Comment by scalarwaves** [comment]
Thank you for this update. Garuda is my new favorite distro. Normally I use the config file to select only the actions I need and skip the rest. Good to know that if I include the `system` action that it will automagically run `garuda-update`.
